konnectivity-client/pkg/client/client_test.go

```diff
@@ -51,14 +51,63 @@ func TestDial(t *testing.T) {
 	}
 
 	if ts.packets[0].Type != client.PacketType_DIAL_REQ {
-		t.Fatalf("expect packet.type %v; got %v", client.PacketType_CLOSE_REQ, ts.packets[0].Type)
+		t.Fatalf("expect packet.type %v; got %v", client.PacketType_DIAL_REQ, ts.packets[0].Type)
 	}
 
 	if ts.packets[0].GetDialRequest().Address != "127.0.0.1:80" {
 		t.Errorf("expect packet.address %v; got %v", "127.0.0.1:80", ts.packets[0].GetDialRequest().Address)
 	}
 }
 
+// TestDialRace exercises the scenario where serve() observes and handles DIAL_RSP
+// before DialContext() does any work after sending the DIAL_REQ.
+func TestDialRace(t *testing.T) {
+	ctx := context.Background()
+	s, ps := pipe()
+	ts := testServer(ps, 100)
+
+	defer ps.Close()
+	defer s.Close()
+
+	tunnel := &grpcTunnel{
+		// artificially delay after calling Send, ensure handoff of result from serve to DialContext still works
+		stream:      fakeSlowSend{s},
+		pendingDial: make(map[int64]pendingDial),
+		conns:       make(map[int64]*conn),
+	}
+
+	go tunnel.serve(&fakeConn{})
+	go ts.serve()
+
+	_, err := tunnel.DialContext(ctx, "tcp", "127.0.0.1:80")
+	if err != nil {
+		t.Fatalf("expect nil; got %v", err)
+	}
+
+	if ts.packets[0].Type != client.PacketType_DIAL_REQ {
+		t.Fatalf("expect packet.type %v; got %v", client.PacketType_DIAL_REQ, ts.packets[0].Type)
+	}
+
+	if ts.packets[0].GetDialRequest().Address != "127.0.0.1:80" {
+		t.Errorf("expect packet.address %v; got %v", "127.0.0.1:80", ts.packets[0].GetDialRequest().Address)
+	}
+}
+
+// fakeSlowSend wraps ProxyService_ProxyClient and adds an artificial 1 second delay after calling Send
+type fakeSlowSend struct {
+	client.ProxyService_ProxyClient
+}
+
+func (s fakeSlowSend) Send(p *client.Packet) error {
+	// send the request so it can start being processed immediately
+	err := s.ProxyService_ProxyClient.Send(p)
+	// delay returning to simulate slowness on the client side,
+	// to exercise serve() observing/handling the DIAL_RSP before
+	// the client does any post-Send() work
+	time.Sleep(time.Second)
```

Comment:
@liggitt now that I think about it, this is not very nice. Sleep is not a good way to sync, maybe we can use the channles or a proper sync mechanism to make sure the other side is where we expect before we continue.